### DISCUSSION TITLE
Pause delayed events?

### DISCUSSION THREAD

**Discussion by screendriver** [statelyai/xstate#2451] [Q&A] [answered]
I have the use case where I need to send a delayed event but with the ability to pause it. For example _"show something for 5 seconds, but when the user moves his mouse hide it and when the user stops moving his mouse show it again BUT only as long as the 5 seconds are reached **in total**"_. So when it was shown already for 2 seconds and the mouse was moved show it again but only for 3 seconds. 

At the moment I'm using [pause-me](https://www.npmjs.com/package/pause-me) for that. I also figured out that delayed events in `xstate` [can be canceled](https://xstate.js.org/docs/guides/delays.html#delayed-events) but unfortunately not paused. 

How would you solve this requirement with `xstate`?

**Comment by Andarist** [comment, marked as answer]
Note that I've written this in a hurry and I have not tested this properly but based on a quick empirical test - it works OK. If not - it at least presents the overall idea how this could be achieved :) Here you go: https://codesandbox.io/s/fervent-franklin-b1sij?file=/src/App.js

**Reply by screendriver** [reply]
Yeah, I had something similar in my mind 🙂 So I was not completely wrong with my idea. I just thought there would be potentially a built-in functionality for this and if not maybe it would be a good idea to implement this directly in `xstate`.

**Reply by davidkpiano** [reply]
>  I just thought there would be potentially a built-in functionality for this and if not maybe it would be a good idea to implement this directly in xstate.

Yep it's a good, creative solution, and I've been thinking about a generic "clock" service for XState v5 and beyond (including server-side applications).

The way this could work is by invoking that service and sending it events:

```js
// TENTATIVE API
import { createClockService, events as clockEvents } from '@xstate/clock';

const machine = createMachine({
  invoke: {
    src: createClockService(),
    id: 'clock'
  },
  on: {
    START: {
      actions: send(clockEvents.start('timer-a', 5000), { to: 'clock' })
    },
    PAUSE: {
      actions: send(clockEvents.pause('timer-a'), { to: 'clock' })
    },
    [clockEvents.expire('timer-a')]: {
      actions: log('Time is up!')
    }
  }
});
```

**Reply by Andarist** [reply]
@davidkpiano this is quite a neat solution - even at the generic level. It effectively offloads a specific concern to a tiny service, allowing the "caller" machine to manage less while improving the readability.
